torch/distributed/fsdp/flat_param.py

```diff
@@ -520,6 +520,10 @@ def shard_metadata(
             self.flat_param._shard_param_offsets[:],  # type: ignore[attr-defined]
         )
 
+    def _flat_param_to(self, *args, **kwargs):
+        """Wraps an in-place call to ``.to()`` for ``self.flat_param``."""
+        self.flat_param.data = self.flat_param.to(*args, **kwargs)
```

Comment:
why can't we just do `self.flat_param = ...` instead of .data?

Reply:
1. This preserves the existing usage: https://github.com/pytorch/pytorch/blob/1e4383f7563c9bb2e3c8e6989b6853d1d04f652f/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1313
2. IIUC: The `FlatParamHandle` holds a reference to a `FlatParameter`, say `p1`, that is currently registered to the `FlattenParamsWrapper`([code](https://github.com/pytorch/pytorch/blob/ed949e22580f6f76a1476626badbc9d6161b7745/torch/distributed/fsdp/flatten_params_wrapper.py#L98)), so it will show up in the `FullyShardedDataParallel.named_parameters()`. If we do not use `self.flat_param.data = ...` here then this will make the `FlatParamHandle` hold a reference to a new `FlatParameter`, say `p2`, that is not the one registered to the `FlattenParamsWrapper`. We would need to refresh the registration in the `FlattenParamsWrapper` to ensure it points to the current `FlatParameter` `p2` and to allow the previous `FlatParameter` `p1` to be garbage collected.

I think that retiring this kind of `.data` usage is outside the scope of this PR stack.